modelforge/train/training.py

```diff
@@ -489,45 +146,105 @@ def _get_energies(
             "per_molecule_energy_predict": per_molecule_energy_predict,
         }
 
+    def _get_charges(
+        self,
+        batch: BatchData,
+        model_prediction: Dict[str, torch.Tensor],
+    ) -> Dict[str, torch.Tensor]:
+        """
+        Computes the total charge of a molecule from the predicted charges from
+        the model.
+        """
+        nnp_input = batch.nnp_input
+        per_atom_charges_predict = model_prediction[
+            "per_atom_charge"
+        ]  # Shape: [num_atoms]
+
+        # Compute predicted total charge
+        total_charge_predict = (
+            torch.zeros_like(model_prediction["per_molecule_energy"])
+            .scatter_add_(
+                dim=0,
+                index=nnp_input.atomic_subsystem_indices.long(),
+                src=per_atom_charges_predict,
+            )
+            .unsqueeze(-1)
+        )  # Shape: [nr_of_molecules, 1]
+
+        dipole_predict = self._predict_dipole_moment(model_prediction, batch)
+        return {
+            "per_molecule_total_charge_predict": total_charge_predict,
+            "per_molecule_total_charge_true": batch.nnp_input.total_charge,
+            "per_molecule_dipole_predict": dipole_predict,
+            "per_molecule_dipole_true": batch.metadata.dipole_moment,
+        }
+
+    def _predict_dipole_moment(
+        self, model_predictions: Dict[str, torch.Tensor], batch: BatchData
+    ) -> torch.Tensor:
+
+        per_atom_charge = model_predictions["per_atom_charge"]
+        # Compute predicted dipole moment
+        dipole_predict = torch.zeros(
+            (model_predictions["per_molecule_energy"].shape[0], 3),
+        ).scatter_add_(
+            dim=0,
+            index=batch.nnp_input.atomic_subsystem_indices.long().unsqueeze(-1),
+            src=per_atom_charge.unsqueeze(-1) * batch.nnp_input.positions,
```

Comment:
This will only work if our molecule center of mass is at zero, which the datasets do not guarantee.  

We could do this as a preprocessing step when we load up the dataset (basically shifting coordinates such that COM is at the origin, as that likely will be easier than storing COM and then trying to do an operation on the entire batch.  This can likely just be done when we do the energy processing.

Reply:
I added in a preprocessing step to shift molecules to the center of mass, which happens basically at the same time we do the energy rescaling and pair list generation.  I'll note that even for spice2, this only took about 30 seconds, so we might want to consider doing this before training, even if we don't need the dipole moment in the loss function.  I did add in a validator in the training pydantic model that will throw an error if dipole moment appears in the loss function but shifting the center of mass is not set to True. 

Another minor change; at some point the code was switched from ignoring records (i.e., not loading that record)  if a record were missing a property of interest, to throwing an error.  I switched this back to having a warning in the log file.  It looks like  a handful of records in spice 2 (mostly seemed to be systems containing ions)  did not have the sci_dipole calculated (or it the array contained NAN).  I can look at that separately.